src/Functions/FunctionJoinGet.cpp

```diff
@@ -65,44 +83,22 @@ FunctionBaseImplPtr JoinGetOverloadResolver<or_null>::build(const ColumnsWithTyp
 {
     auto [storage_join, attr_name] = getJoin(arguments, context);
     auto join = storage_join->getJoin();
-    DataTypes data_types(arguments.size());
-
+    DataTypes data_types(arguments.size() - 2);
+    for (size_t i = 2; i < arguments.size(); ++i)
+        data_types[i - 2] = arguments[i].type;
+    auto return_type = join->joinGetCheckAndGetReturnType(data_types, attr_name, or_null);
     auto table_lock = storage_join->lockForShare(context.getInitialQueryId(), context.getSettingsRef().lock_acquire_timeout);
-    for (size_t i = 0; i < arguments.size(); ++i)
-        data_types[i] = arguments[i].type;
-
-    auto return_type = join->joinGetReturnType(attr_name, or_null);
     return std::make_unique<FunctionJoinGet<or_null>>(table_lock, storage_join, join, attr_name, data_types, return_type);
 }
 
 template <bool or_null>
-DataTypePtr JoinGetOverloadResolver<or_null>::getReturnType(const ColumnsWithTypeAndName & arguments) const
-{
-    auto [storage_join, attr_name] = getJoin(arguments, context);
-    auto join = storage_join->getJoin();
-    return join->joinGetReturnType(attr_name, or_null);
-}
-
-
-template <bool or_null>
-void ExecutableFunctionJoinGet<or_null>::execute(Block & block, const ColumnNumbers & arguments, size_t result, size_t input_rows_count)
-{
-    auto ctn = block.getByPosition(arguments[2]);
-    if (isColumnConst(*ctn.column))
-        ctn.column = ctn.column->cloneResized(1);
-    ctn.name = ""; // make sure the key name never collide with the join columns
-    Block key_block = {ctn};
-    join->joinGet(key_block, attr_name, or_null);
-    auto & result_ctn = key_block.getByPosition(1);
-    if (isColumnConst(*ctn.column))
-        result_ctn.column = ColumnConst::create(result_ctn.column, input_rows_count);
-    block.getByPosition(result) = result_ctn;
-}
-
-template <bool or_null>
-ExecutableFunctionImplPtr FunctionJoinGet<or_null>::prepare(const Block &, const ColumnNumbers &, size_t) const
+void JoinGetOverloadResolver<or_null>::checkNumberOfArgumentsIfVariadic(size_t number_of_arguments) const
 {
-    return std::make_unique<ExecutableFunctionJoinGet<or_null>>(join, attr_name);
+    if (number_of_arguments < 3)
+        throw Exception(
+            "Number of arguments for function " + getName() + " doesn't match: passed " + toString(number_of_arguments)
+                + ", should be greater or equal to 3",
+            ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
```

Comment:
I don't see where checkNumberOfArgumentsIfVariadic() is called, in this case it should bail out in getJoin when it access arguments w/o checking, but why the tests passed?

Interesting, it does not SIGSEGV's because there is std::out_of_range exception:

>2020.07.15 06:50:15.428892 [ 225276 ] {a1cfa369-a232-4df9-a456-95ee0af99391} <Debug> executeQuery: (from [::1]:44804) SELECT joinGet(); 
>2020.07.15 06:50:15.429338 [ 225276 ] {a1cfa369-a232-4df9-a456-95ee0af99391} <Error> executeQuery: std::exception. Code: 1001, type: std::out_of_range, e.what() = vector (version 20.7.1.4090) (from [::1]:44804) (in query: SELECT joinGet(); )

@akuzm worth converting out_of_range to `abort()` ? (like in #12522 and related)

Reply:
FWIW this is the only std exception in the log:

```shell
$ fgrep -a 'type: std::' clickhouse-server.log  -c
99
```

(log from this build stateless tests)

Reply:
>I don't see where checkNumberOfArgumentsIfVariadic() is called, in this case it should bail out in getJoin when it access arguments w/o checking

#12705

>@akuzm worth converting out_of_range to abort() ? (like in #12522 and related)

#12704